### PR TITLE
[WIP] Make structs parametric and type stable and take constructors out

### PR DESCRIPTION
**Pull request by mohamed82008**
Closes https://github.com/TuringLang/Turing.jl/issues/486. I took the constructors out because that's easier when dealing with parameters since the non-parametric constructor is defined by default and I can call it from other external constructors.

### CODE REVIEW

File: src/core/io.jl [hunk 1]
```diff
@@ -86,18 +86,18 @@ mean(chain[:mu])      # find the mean of :mu
 mean(chain[:sigma])   # find the mean of :sigma
 ```
 """
-mutable struct Chain <: AbstractChains
+mutable struct Chain{R<:AbstractRange{Int}} <: AbstractChains
```
Comment (willtebbutt) [inline src/core/io.jl]: Consider `AbstractRange{<:Integer}`, or perhaps simple `AbstractRange`.

File: src/samplers/pmmh.jl [hunk 2]
```diff
@@ -11,17 +11,17 @@ alg = PMMH(100, SMC(20, :v1), MH(1,:v2))
 alg = PMMH(100, SMC(20, :v1), MH(1,(:v2, (x) -> Normal(x, 1))))
 ```
 """
-mutable struct PMMH <: InferenceAlgorithm
+mutable struct PMMH{T, A} <: InferenceAlgorithm
```
Comment (willtebbutt) [inline src/samplers/pmmh.jl]: Consider `A <: Tuple`, to ensure the same type constraints are in place as before. I'm not sure if this is entirely necessary though.

**Comment by willtebbutt** [comment]
How far are you planning on pushing these changes? There's still quite a lot of concretely typed fields (eg. in `Chain`) that it might be better to make parametric. I appreciate that this is a WIP, so maybe you're already thinking about this...

File: src/core/io.jl (hunk 1 above)
Comment (mohamed82008) [reply src/core/io.jl]: Any reason?

File: src/samplers/pmmh.jl (hunk 2 above)
Comment (mohamed82008) [reply src/samplers/pmmh.jl]: Sure.

**Comment by mohamed82008** [comment]
I tried using a specific range type in `Chain` but apparently it changes somewhere in the code so that gave a conversion error. I will look into it again.

**Comment by mohamed82008** [comment]
I guess the rule of thumb is when you need more parameters, add them. Parameters come at the cost of convenience sometimes. This PR just deals with type instability issues, not arbitrary types of fields, e.g. handling `Float32`s. This can be done in a different PR after the design of the package has stabilized.

**Comment by yebai** [comment]
> I tried using a specific range type in Chain but apparently it changes somewhere in the code so that gave a conversion error. I will look into it again.

Martin is working towards a rewrite of `io.jl` -- let's leave `Chain` as is for now.

File: src/core/io.jl (hunk 1 above)
Comment (willtebbutt) [reply src/core/io.jl]: It's just less concrete, and I don't think we really lose anything. A completely generic `AbstractRange` might be a bit loose though actually.

**Comment by mohamed82008** [comment]
So I rebased this on master. If tests pass and no further comments are there, I believe this is ready to go in.

**Comment by yebai** [comment]
Thanks @mohamed82008!

**Pull request merged**